### DISCUSSION TITLE
DatePicker API unique

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#38438] [General]
## Motivation

Unique DatePicker related API. Deprecated part of it and merge into single API. Consider perf saving.

## Related Issue

* https://github.com/ant-design/ant-design/issues/33795
* https://github.com/ant-design/ant-design/issues/33922
* https://github.com/ant-design/ant-design/issues/21635

#### Unique: `dateRender`, & `monthCellRender`

To `cellRender(date, info: { type: 'date' | 'month' | 'year', range, locale })`

#### `disabledDate` support preset range

```tsx
<DatePicker
  disabledDate={[
    { min: dayjs(xxx), max: dayjs(yyy) },
    { min: dayjs(aaa), max: dayjs(bbb) },
  ]}
/>
```

### Disabled Tips

![跨时间范围内的不可选日期-及时阻拦](https://user-images.githubusercontent.com/5378891/75410784-f5422100-5957-11ea-9ac5-c31846762161.jpg)

**Comment by afc163** [comment]
https://github.com/ant-design/ant-design/issues/21635

这个还搞不

**Reply by zombieJ** [reply]
及时阻拦感觉不错，搞起~

**Comment by afc163** [comment]
应该是取并集？

**Reply by zombieJ** [reply]
嗯

**Comment by afc163** [comment]
如果要有动态判断配合怎么办，是不是 disabledDate 里面也支持 function？

**Reply by zombieJ** [reply]
动态的话还是直接 function，这个 preset range 就是为了性能考虑不做成动态的。

**Reply by afc163** [reply]
min max 有可能和 function 同时使用的，不是互斥的，现在 MUI 里两个 API 也是分开的。

**Reply by zombieJ** [reply]
那这样？
```tsx
<DatePicker
  disabledDate={[
    { min: dayjs(xxx), max: dayjs(yyy) },
    { min: dayjs(aaa), max: dayjs(bbb) },
    function(date) {},
  ]}
/>
```

**Comment by afc163** [comment]
需要有 disabledMonth 和 disabledYear 等么

**Reply by zombieJ** [reply]
disabledMonth, disabledYear 可以用上面的 preset 搞定。单独开接口会比较奇怪，因为如果用户在 disabledDate 里禁用了，disabledYear 没有禁用。那其实代码逻辑还是不得不运行一次 disabledDate 来确保是否可禁用。

**Reply by afc163** [reply]
举个代码例子？

**Reply by zombieJ** [reply]
```tsx
// 伪代码
<DatePicker
  disabledDate={(date) => date < '2022-01-01' || date > '2022-12-31'}
  disabledYear={(date) => date < '2000' || date > '2300'}
/>
```

**Reply by afc163** [reply]
所以还是要单开？

**Reply by zombieJ** [reply]
不单开，就是上面提到的通过预设范围来实现。disabledDate 是唯一且统一的 API
